### PR TITLE
feat: include local copy of v4l2 headers

### PR DESCRIPTION
**Pull request by ajcasagrande**
Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

image

### CODE REVIEW

**Comment by vladimirvivien** [comment]
Thank you @ajcasagrande . This looks like a great contribution and will be reviewing it this week.

File: include/README.md
Comment (vladimirvivien) [inline include/README.md:1]: This is awesome !!!
Maybe in the future we explore automated pull via Git sumodules + GitHub dependabot.

**Review by vladimirvivien — APPROVED** [review]
@ajcasagrande This a great contribution!!!! Thank you.

**Comment by farshidtz** [comment]
I don't think this makes any sense. The build should be done against the same kernel headers that the application is expected to run on. Otherwise, there could be links to wrong addresses and files that exist during the build but not on runtime.

**Comment by ajcasagrande** [comment]
@farshidtz A few things here:

1. I did not originally add this text to the PR comment, but it is included [in the files I have committed](https://github.com/vladimirvivien/go4vl/pull/20/files#diff-aca3b3a4fd81ce3e9757afee4ddd97a69d4cf83f36bef3617a90380f696c1fef) (emphasis mine):

> Here we provide a copy of the v4l2 header files as per the 
> [kernel.org docs](https://docs.kernel.org/userspace-api/media/v4l/func-ioctl.html#description):
> 
> > Macros and defines specifying V4L2 ioctl requests are located in the `videodev2.h` header file.
> > _**Applications should use their own copy, not include the version in the kernel sources on the system they compile on.**_
>
> This helps to reduce compilation errors due to outdated headers installed in the user's system.

2.  The alternative recommended way to fix compilation issues was to sideload a newer version of the entire linux headers onto your system. That seems like a worse recommendation than us including the needed headers in the source code here.

3. A majority of these headers are `#define` constants that will eventually be hard-coded values thanks to the pre-processor. The rest of them are enums and struct definitions which I assume that the v4l2 devs try and keep backward and forward compatible.

**Comment by farshidtz** [comment]
I'm not an expert in this area but I suggest doing further research to ensure a successful build leads to an application that doesn't fail unexpectedly in a production environment.

> 2\. The alternative recommended way to fix compilation issues was to sideload a newer version of the entire linux headers onto your system. That seems like a worse recommendation than us including the needed headers in the source code here.

This is not ideal either. Adding the headers will just makes the build pass but things will fail if a new feature is used without having the actual implementation present.

> 3\. A majority of these headers are `#define` constants that will eventually be hard-coded values thanks to the pre-processor. The rest of them are enums and struct definitions which I assume that the v4l2 devs try and keep backward and forward compatible.

Could you provide any reference to support this claim? I am looking for a policy set to prevent changing the macro (e.g. `#define V4L2_SEL_TGT_COMPOSE_PADDED	0x0103`) values or changes in enums and other definitions in future.

See for example this statement for the Kernel Build System:
> Kernel headers are backwards compatible, but not forwards compatible. This means that a program built against a C library using older kernel headers should run on a newer kernel (although it may not have access to new features), but a program built against newer kernel headers may not work on an older kernel.
https://www.kernel.org/doc/html/v5.15/kbuild/headers_install.html

**Comment by ajcasagrande** [comment]
I'm not an expert on this either, however, the [official documentation](https://linuxtv.org/downloads/v4l-dvb-apis-new/userspace-api/v4l/func-ioctl.html#description) say we should include the headers, and the popular `v4l-utils` apps [include them as well](https://git.linuxtv.org/v4l-utils.git/tree/include).

I have searched around for a good reasoning from somewhere on _why_ this is the case, and unfortunately I have not found one. However the above reasonings lead me to believe it is a good practice to include them here.

**Comment by farshidtz** [comment]
> I'm not an expert on this either, however, the [official documentation](https://linuxtv.org/downloads/v4l-dvb-apis-new/userspace-api/v4l/func-ioctl.html#description) say we should include the headers, and the popular `v4l-utils` apps [include them as well](https://git.linuxtv.org/v4l-utils.git/tree/include).
> 

The sentence is "Applications should use their own copy, not include the version in the kernel sources on the system they compile on.". But this is a library, not an application. The applications, using the library, should use their own copy, maintain and document it accordingly.

> I have searched around for a good reasoning from somewhere on why this is the case, and unfortunately I have not found one. However the above reasonings lead me to believe it is a good practice to include them here.

IMO, if the headers are to remain, the library should:
- Specify the Kernel version from which the headers are taken from and leave the burden of identifying backward/forward compatibility to users.
- Add a statement clarifying if the headers will be updated by the maintainers or remain hardcoded and eventually go out of date.
- Ensure that its license remains correct. The SPDX-License-Identifier for the included headers are  `((GPL-2.0+ WITH Linux-syscall-note) OR BSD-3-Clause)` but this library has MIT License.

**Comment by ajcasagrande** [comment]
For the last part:

From what I have seen, the BSD-3-Clause is compatible with MIT and as far as I know BSD licensed code can be included in MIT projects as long as the license headers remain intact. Also dual-licensed code means you can pick which to use, which in this case the BSD-3 would be the preferred one.

**Comment by vladimirvivien** [comment]
I appreciate this discussion very much. I thank you for your input @farshidtz and @ajcasagrande .  After this got merged, I started thinking about the licensing issue. Also, I agree that if this is compiled against a specific commit, it should be prominently displayed/documented.

Meanwhile, I am doing some additional research to see how other Go projects handle this.  What I have seen so far
*  Project that includes the headers, as is done here
* Project that automatically pull a specific version at commit, via submodules

I think there is a middle ground and will continue to research how best to do this without putting any burden on users of this library. I am thinking about having a small codegen package that automates pulling a specific version and prepare the cgo stubs. That way, users can select a kernel version that works for them (and removing the licensing concern).

Let me know what you both think. Thanks again.

**Pull request merged**
